kv: decompose TxnCoordSender into a stack of txnReqInterceptors

nvb opened the pull request:
The change introduces the notion of a `txnReqInterceptor` interface. Instances
of the interface are pluggable request interceptors that transform requests and
responses and can perform specific retries in the context of a transaction.
The general idea is that they fill the role of a `Sender` stack that wants to
be slightly more tightly coupled to `TxnCoordSender`. This coupling takes the
form of sane unified locking and some shared state.

There were two main reasons for performing this refactoring now:
1. transactional pipelining (#16026) will require adding even more
   state to TxnCoordSender. Specifically, TxnCoordSender will now
   need to track all outstanding writes and ensure that any overlapping
   requests use `QueryIntent` to chain on to these writes. These outstanding
   writes will be tracked in TxnCoordMeta. This all fits very nicely into
   the framework of txnReqInterceptors, allowing the new logic to be completely
   abstracted away from the mess that is TxnCoordSender into a `txnPipeliner`
   interceptor.
2. it abstracts most of TxnCoordSender's logic into smaller plugable components.
   After creating the initial set of txnReqInterceptors, the only thing left in
   TxnCoordSender is its TxnCoordMeta, its ability to asyncronously abort
   transaction, and its heartbeat goroutine. This massively reduces the burden
   of getting rid of TxnCoordSender entirely and lifting everything it does up
   into `client.Txn`. It also opens up the opportunity for more directed testing
   of this logic (wishful thinking?). For instance, instead of testing that an
   EndTxn cleans up intents, we could test that a `txnIntentCollector` correctly
   adds the tracked intents onto the EndTxn req during its `beforeSendLocked`
   transformation.

Release note: None

nvb committed:
kv: decompose TxnCoordSender into a stack of txnReqInterceptors

NOTE: I don't necessarily think this should be merged as-is, but I'd like
to get opinions on how people feel about its approach. The problem it is
trying to solve is that TxnCoordSender performs a large number of different
operations/transformations and maintains a lot of random state in a very
unstructured manner. It has very few abstraction boundaries and it's unclear
how to add any new functionality to it without touching and risking breaking
everything else. Even worse, it's not clear how to test new functionality
in the TxnCoordSender from less than a mile away.

The change introduces the notion of a `txnReqInterceptor` interface. Instances
of the interface are pluggable request interceptors that transform requests and
responses and can perform specific retries in the context of a transaction.
The general idea is that they fill the role of a `Sender` stack that wants to
be slightly more tightly coupled to `TxnCoordSender`. This coupling takes the
form of sane unified locking and some shared state.

There were two main reasons for performing this refactoring now:
1. transactional pipelining (#16026) will require adding even more
   state to TxnCoordSender. Specifically, TxnCoordSender will now
   need to track all outstanding writes and ensure that any overlapping
   requests use `QueryIntent` to chain on to these writes. These outstanding
   writes will be tracked in TxnCoordMeta. This all fits very nicely into
   the framework of txnReqInterceptors, allowing the new logic to be completely
   abstracted away from the mess that is TxnCoordSender into a `txnPipeliner`
   interceptor.
2. it abstracts most of TxnCoordSender's logic into smaller pluggable components.
   After creating the initial set of txnReqInterceptors, the only thing left in
   TxnCoordSender is its TxnCoordMeta, its ability to asyncronously abort
   transaction, and its heartbeat goroutine. This massively reduces the burden
   of getting rid of TxnCoordSender entirely and lifting everything it does up
   into `client.Txn`. It also opens up the opportunity for more directed testing
   of this logic (wishful thinking?). For instance, instead of testing that an
   EndTxn cleans up intents, we could test that a txnIntentCollector correctly
   adds the tracked intents onto the EndTxn req during its `beforeSendLocked`
   transformation.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/26496)
<!-- Reviewable:end -->

spencerkimball commented:
I have strong objections to the stated intent (in the PR description, not the PR) to remove `TxnCoordSender` and lift everything into `client.Txn`. I do believe these two things should be merged, but the `client` package should not be the place where that's done. That package is otherwise a shim and was never meant to contain complex logic, especially not logic which is concerned with things like intents and refresh spans, etc. Just want to get that on the record.

This LGTM; I'm assuming most of the deleted and relocated code is verbatim. Please let me know if there are areas which require closer inspection, because I didn't compare line by line.

---

Review status: 0 of 5 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---

*[pkg/kv/txn_coord_sender.go, line 135 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEPhO-V42neZ7d4J8sM:-LEPhO-V42neZ7d4J8sN:b-h9qrbj) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a4afa35e05648591aba772f2fc240d4e61b56f0/pkg/kv/txn_coord_sender.go#L135)):*
> ```Go
> 	) (roachpb.BatchRequest, *roachpb.Error)
> 
> 	// maybeRetrySend inspects the esponse batch and error and optionally
> ```

s/esponse/response/

---

*[pkg/kv/txn_coord_sender.go, line 282 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEPebBT2tkChHB-HFB_:-LEPebBT2tkChHB-HFBa:b-r9ngza) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a4afa35e05648591aba772f2fc240d4e61b56f0/pkg/kv/txn_coord_sender.go#L282)):*
> ```Go
> 
> 	// Create a stack of request/response interceptors.
> 	var ri *RangeIterator
> ```

This seems pretty heavyweight from an allocation perspective since it's done on every request.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-:-LEPdQoa5pCKZ_JkKTSD:b-3eaubs)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
I heart the PR

---

Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/kv/txn_batch_wrapper.go, line 25 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEQMTRfEiGCleWq5VfP:-LEQMTRfEiGCleWq5VfQ:bgeupab) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a4afa35e05648591aba772f2fc240d4e61b56f0/pkg/kv/txn_batch_wrapper.go#L25)):*
> ```Go
> )
> 
> // txnBatchWrapper is an implementation of txnReqInterceptor. It catches
> ```

nit: `txnBatchWrapper is a txnReqInterceptor that catches ..."

---

*[pkg/kv/txn_batch_wrapper.go, line 33 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEQKPVs-NvxzlourdLC:-LEQKPVs-NvxzlourdLD:b64vck4) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a4afa35e05648591aba772f2fc240d4e61b56f0/pkg/kv/txn_batch_wrapper.go#L33)):*
> ```Go
> // run the whole thing for them, or any restart will still end up at the client
> // which will not be prepared to be handed a Txn.
> // TODO(andreimatei): we should no longer send non-transaction requests
> ```

s/andreimatei/andrei

The person's name inside a TODO is defined to be the person that others can ask about the note. If you want that to be me, please make the note clearer :)
"we should no longer send" - you mean "if we lifted the retry loop, then we wouldn't have to send through the TCS", right?

---

*[pkg/kv/txn_batch_wrapper.go, line 42 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEQMt1x59PXZLfhjYNZ:-LEQMt1x59PXZLfhjYN_:bnf6s5n) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a4afa35e05648591aba772f2fc240d4e61b56f0/pkg/kv/txn_batch_wrapper.go#L42)):*
> ```Go
> var _ txnReqInterceptor = &txnBatchWrapper{}
> 
> func (*txnBatchWrapper) beforeSendLocked(
> ```

pretty please annotate all methods with `is part of txnReqInterceptor`

---

*[pkg/kv/txn_batch_wrapper.go, line 58 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEQMhoA6mPUdu0xUdFJ:-LEQMhoA6mPUdu0xUdFK:bgf3kx0) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a4afa35e05648591aba772f2fc240d4e61b56f0/pkg/kv/txn_batch_wrapper.go#L58)):*
> ```Go
> 	// Run a one-off transaction with that single command.
> 	if log.V(1) {
> 		log.Infof(ctx, "%s: auto-wrapping in txn and re-executing: ", ba)
> ```

nit: `log.VEventf(ctx, 2)`

---

*[pkg/kv/txn_coord_sender.go, line 284 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEQOMxH0NRVY6xnCCn_:-LEQOMxH0NRVY6xnCCna:blboqud) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a4afa35e05648591aba772f2fc240d4e61b56f0/pkg/kv/txn_coord_sender.go#L284)):*
> ```Go
> 	var ri *RangeIterator
> 	if ds, ok := tcf.wrapped.(*DistSender); ok {
> 		ri = NewRangeIterator(ds)
> ```

why did you move the creation of the `RangeIterator` here? Why doesn't the `txnIntentCollector` do it internally?

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-:-LEQOk5LAyIO61eYIWgp:b-6dvgux)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Oh, but please do me a big favor and rebase on #23055.  I want to merge that today if you give me a few reviews :)

---

Review status: 0 of 5 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-:-LEQOxYq7dCzpF1xqRst:b-2dlhht)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: 0 of 5 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/kv/txn_coord_sender.go, line 282 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEPebBT2tkChHB-HFB_:-LER9cdfBvQY_YpOtn8R:bnmkw96) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a4afa35e05648591aba772f2fc240d4e61b56f0/pkg/kv/txn_coord_sender.go#L282)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This is only done once per transaction epoch. Is that what you meant by "every request"?
</blockquote></details>

Yes. I’m always thinking about 1PC single inserts from a perf perspective. You tell me if this is going to be problematic.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26496)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Review status: 0 of 5 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/kv/txn_coord_sender.go, line 282 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEPebBT2tkChHB-HFB_:-LERDfoQ-1ma3p1WGgqL:brc5z75) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a4afa35e05648591aba772f2fc240d4e61b56f0/pkg/kv/txn_coord_sender.go#L282)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

Yes. I’m always thinking about 1PC single inserts from a perf perspective. You tell me if this is going to be problematic.
</blockquote></details>

I doubt it, but it's a valid concern to have. An easy fix is to add an instance of each of these interceptor structs to `TxnCoordSender` so that everything gets allocated at once. I'll go ahead and do that.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26496)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 0 of 5 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/kv/txn_coord_sender.go, line 284 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEQOMxH0NRVY6xnCCn_:-LEWHli69AgycTQhUuPs:bhz5mve) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a4afa35e05648591aba772f2fc240d4e61b56f0/pkg/kv/txn_coord_sender.go#L284)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I lifted it out here to limit the scope of what we give to `txnIntentCollector`. It shouldn't need a `Sender` or a `DistSender`, it just needs a `RangeIterator`. The less we give these interceptors, the easier future refactoring will be and the easier they are to test.
</blockquote></details>

pls add a comment to the `txnIntentCollecter` explaining that it's being configured through this `ri` field.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26496)*
<!-- Sent from Reviewable.io -->

nvb committed:
kv: stop sharing state between txnReqInterceptors and TxnCoordSender

Before this change, each txnReqInterceptor held a reference to its parent
TxnCoordSender's TxnCoordMeta. Each interceptor would update its individual
part of the meta separately (under lock), but a TxnCoordSender could always
reach into its own meta and perform any modifications it wanted. As such,
there wasn't a hard abstraction boundary. An example of how this abstraction
was broken was `tryAsyncAbort`, which used to pull the intents out from under
a txnIntentCollector.

This change removes this shared state. `txnReqInterceptors` now hold on to
their own state and this is only accessible through new `populateMetaLocked`
and `augmentMetaLocked` methods. Because of this change, `TxnCoordSender`
no longer has any interest or knowledge about intents or refresh spans.
These concepts are all completely contained in their respective interceptors.

The change also moves away from holding all state in a `TxnCoordMeta` and
allowing this state to be pulled out at any time. This is an important
change as we look towards the `txnPipeliner` interceptor because that will
hold more complex data structures of in-flight writes. These writes will
need to be serialized into a `TxnCoordMeta` when creating LeafTxn coordinators,
but this should be on-demand (see `populateMetaLocked`). A similar change
can now be made in `txnIntentCollector` along the lines of this
[suggestion](https://github.com/cockroachdb/cockroach/pull/21078#issuecomment-356654541).
I added a TODO to explore.

This change also simplifies the handling of concurrent requests and their
effect on refresh spans. In doing so, it revealed a race where the meta's
Txn.RefreshedTimestamp was only updated under lock before performing the span
refresh if a prefix of a batch was executed. If not (`if br == nil {` in
`maybeRetrySend`) then spans could be "missed" by the refresh. This was
subtle so this commit fixed the bug and made the race prevention more explicit.

Release note: None

nvb committed:
kv: allocate txnReqInterceptors and TxnCoordSender together

Before this change, each txnReqInterceptor was allocated individually
and a new slice was allocated to hold all txnReqInterceptors. This
change adjusts this so that all txnReqInterceptors are allocated together
with TxnCoordSender, along with an array to store them all.

Release note: None

nvanbenschoten commented:
I'm reworking this into a stack of `Sender`-like components that are called under lock with a bottom-level sender that unlocks the `TxnCoordSender` before passing a request to `DistSender` (or whatever the next wrapped sender is). This provides that same nice locking properties introduced here, allowing for better modularization, while simplifying/improving mid-stack retry handling and cleaning up the `txnReqInterceptor`. This has the nice side-effect of making the refresh spans optimization more powerful because we can now configure how many times a single request can refresh its spans before giving up, instead of only trying once because that's all the code structure allowed for.

nvanbenschoten commented:
... so please don't spend time reviewing quite yet.

nvb committed:
kv: restructure txnInterceptors into a recursive lockedSender stack

This change restructures `txnInterceptors` into a stack of `lockedSenders`.
`lockedSenders` are like normal `client.Senders` except that their contract
enforces that callers hold the `TxnCoordSender's` lock when calling their
single `SendLocked` method. This provides the desired locking property
introduced in an earlier earlier commit, allowing for improved modularization,
while simplifying/improving mid-stack retry handling. Retries are improved
because `txnInterceptors` now maintain a call-stack across the sending and
receiving of a given request. This also allows the `txnInterceptors` interface
to be simplified and for any state request-specific state (e.g. `needBeginTxn`)
to be easily preserved across the RPC call.

Permitting this to all work is a new `txnLockGatekeeper` `lockedSender`,
which is attached to the bottom of the `txnInterceptor` stack. This type
is responsible for unlocking the coordinator's mutex when sending a request
and locking the coordinator's mutex when receiving a response.

This change also has the effect of dramatically cleaning up the locking in
`TxnCoordSender` itself. In `Send` it simply locks, preforms some prep,
calls `SendLocked`, and continues on with the lock after the response.

Release note: None

nvanbenschoten commented:
This has been updated with the change alluded to earlier and rebased on master. PTAL.

tschottdorf commented:
Looks pretty good to me! Still lots of complexity (and lots of code movement, necessarily so hard to review exhaustively) but :thumbsup: on the general approach. Thank you for doing this.

---

Reviewed 1 of 5 files at r1, 5 of 5 files at r5, 1 of 1 files at r6, 5 of 5 files at r7, 5 of 5 files at r8.
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/kv/txn_batch_wrapper.go, line 1 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEy9xZ16KtcN-Hhmkc0:-LEy9xZ16KtcN-Hhmkc1:bf7huyu) ([raw file](https://github.com/cockroachdb/cockroach/blob/c00da02874a2d2071a8f365974045fd9230f042c/pkg/kv/txn_batch_wrapper.go#L1)):*
> ```Go
> // Copyright 2018 The Cockroach Authors.
> ```

Maybe the scheme `(txn_)?interceptor_<name>.go` is more appropriate. Your call.

---

*[pkg/kv/txn_coord_sender.go, line 171 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEy3b6UDxPGAUTjos-T:-LEy3b6UDxPGAUTjos-U:b97cvdl) ([raw file](https://github.com/cockroachdb/cockroach/blob/c00da02874a2d2071a8f365974045fd9230f042c/pkg/kv/txn_coord_sender.go#L171)):*
> ```Go
> 	// refreshMetaLocked refreshes any internal state held inside the
> 	// interceptor that is a function of the TxnCoordMeta. It is called
> 	// when the TxnCoordMeta updates its TxnCoordMeta manually.
> ```

"when" -> before or after?

---

*[pkg/kv/txn_coord_sender.go, line 491 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEy4COI7CfiIxYp6ILB:-LEy4COI7CfiIxYp6ILC:bb9p8iq) ([raw file](https://github.com/cockroachdb/cockroach/blob/c00da02874a2d2071a8f365974045fd9230f042c/pkg/kv/txn_coord_sender.go#L491)):*
> ```Go
> 				ba, pErr = reqInt.beforeSendLocked(ctx, ba)
> 				if pErr != nil {
> 					return ba, pErr
> ```

why return `ba` on error here and above? Either something subtle is happening or it should be changed. But it looks like the caller just throws `ba` away on error.

---

*[pkg/kv/txn_coord_sender.go, line 518 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEy4_6e8Q4evqEXiyWd:-LEy4_6e8Q4evqEXiyWe:b-xksbmk) ([raw file](https://github.com/cockroachdb/cockroach/blob/c00da02874a2d2071a8f365974045fd9230f042c/pkg/kv/txn_coord_sender.go#L518)):*
> ```Go
> 			for i := len(tc.interceptors) - 1; i >= 0; i-- {
> 				reqInt := tc.interceptors[i]
> 				br, pErr = reqInt.maybeRetrySend(ctx, &ba, br, pErr)
> ```

The abstraction is a little leaky here, right? On the way out, the interceptor sees the request before any later interceptors made their modifications. Then, on the way back, it gets a `ba` that has seen all of the modifications and in particular may have been altered by all later senders both on the way out and in.

Not necessarily a problem, but call this out in the command on `maybeRetrySend`.

---

*[pkg/kv/txn_coord_sender.go, line 186 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEyAn646UtjckUjk4lt:-LEyAn653p9YC3j3BWPH:b-80ul4c) ([raw file](https://github.com/cockroachdb/cockroach/blob/93303e5ae835bb45edf1d1310209c906c0f7ef52/pkg/kv/txn_coord_sender.go#L186)):*
> ```Go
> 	populateMetaLocked(meta *roachpb.TxnCoordMeta)
> 
> 	// refreshMetaLocked refreshes any internal state held inside the
> ```

Stale

---

*[pkg/kv/txn_coord_sender.go, line 334 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEyTmuG8BIvJ-6pFfIK:-LEyTmuG8BIvJ-6pFfIL:bwhg76w) ([raw file](https://github.com/cockroachdb/cockroach/blob/93303e5ae835bb45edf1d1310209c906c0f7ef52/pkg/kv/txn_coord_sender.go#L334)):*
> ```Go
> 	tcs.interceptorAlloc.txnSpanRefresher = txnSpanRefresher{
> 		st:           tcf.st,
> 		mu:           &tcs.mu,
> ```

:thumbsup: there goes my earlier comment.

---

*[pkg/kv/txn_coord_sender.go, line 213 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEyTgyU-gU3bcLdU8ei:-LEyTgyV-QXODggh4OPX:b-flsg3) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef672433a8a1af1767d681d699bd57b817f059a7/pkg/kv/txn_coord_sender.go#L213)):*
> ```Go
> 	ctx context.Context, ba roachpb.BatchRequest,
> ) (*roachpb.BatchResponse, *roachpb.Error) {
> 	gs.mu.Unlock()
> ```

Sneaky, I like it. But you could put a comment here.

---

*[pkg/kv/txn_intent_collector.go, line 72 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEy649u2bEHcRB2ng_Z:-LEy649vDoaMBuXDcyEQ:b-6y2yjg) ([raw file](https://github.com/cockroachdb/cockroach/blob/c00da02874a2d2071a8f365974045fd9230f042c/pkg/kv/txn_intent_collector.go#L72)):*
> ```Go
> 	et := rArgs.(*roachpb.EndTransactionRequest)
> 	if len(et.IntentSpans) > 0 {
> 		// TODO(tschottdorf): it may be useful to allow this later.
> ```

I wouldn't mind if this TODO had a ... happy little accident :gun: 

---

*[pkg/kv/txn_intent_collector.go, line 133 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEy7ZYg5asXjxjZnAED:-LEy7ZYg5asXjxjZnAEE:b-yvct2d) ([raw file](https://github.com/cockroachdb/cockroach/blob/c00da02874a2d2071a8f365974045fd9230f042c/pkg/kv/txn_intent_collector.go#L133)):*
> ```Go
> ) (*roachpb.BatchResponse, *roachpb.Error) {
> 	// Append the intents from the current request to our txn meta.
> 	if err := ic.appendAndCondenseIntentsLocked(ctx, ba, br); err != nil {
> ```

I haven't peeked inside of this method, but it seems pretty bad if we're sometimes missing intent spans, for that could lead to lost writes if the txn ends up committing.

---

*[pkg/kv/txn_intent_collector.go, line 140 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEy7jYdEdjhYmu54R7a:-LEy7jYdEdjhYmu54R7b:b-fg20o8) ([raw file](https://github.com/cockroachdb/cockroach/blob/c00da02874a2d2071a8f365974045fd9230f042c/pkg/kv/txn_intent_collector.go#L140)):*
> ```Go
> 
> // refreshMetaLocked implements the txnReqInterceptor interface.
> func (ic *txnIntentCollector) refreshMetaLocked() {
> ```

Maybe this method should be called `metaRefreshedLocked()`?

---

*[pkg/kv/txn_intent_collector.go, line 157 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEy7ySz8m0BxyO66Lck:-LEy7ySz8m0BxyO66Lcl:bsbfaif) ([raw file](https://github.com/cockroachdb/cockroach/blob/c00da02874a2d2071a8f365974045fd9230f042c/pkg/kv/txn_intent_collector.go#L157)):*
> ```Go
> 		ic.intentsSizeBytes += int64(len(span.Key) + len(span.EndKey))
> 	})
> 	condensedIntents, condensedIntentsSize, err := ic.maybeCondenseIntentSpans(
> ```

The mystery continues.

---

*[pkg/kv/txn_intent_collector.go, line 177 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEy86NqBpyE8aA3hCM6:-LEy86NqBpyE8aA3hCM7:b-ssy11r) ([raw file](https://github.com/cockroachdb/cockroach/blob/c00da02874a2d2071a8f365974045fd9230f042c/pkg/kv/txn_intent_collector.go#L177)):*
> ```Go
> // in the txn intent collector itself. Spans are condensed based on
> // current range boundaries. Returns the condensed set of spans and
> // the new total spans size. Note that errors can be returned if the
> ```

Aha! I don't think that should return an error. It should just leave the spans uncondensed.

---

*[pkg/kv/txn_intent_collector.go, line 116 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEyUKQE4xfcacFZcE-Z:-LEyUKQE4xfcacFZcE-_:b-jm1zx7) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef672433a8a1af1767d681d699bd57b817f059a7/pkg/kv/txn_intent_collector.go#L116)):*
> ```Go
> 	}
> 
> 	// Send through wrapped lockedSender. Unlocks while sending then re-locks.
> ```

What's the comment for? This method doesn't care any more, right?

---

*[pkg/kv/txn_intent_collector.go, line 126 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEyUQYe2iKY1XbJL4SY:-LEyUQYe2iKY1XbJL4SZ:bagwqnn) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef672433a8a1af1767d681d699bd57b817f059a7/pkg/kv/txn_intent_collector.go#L126)):*
> ```Go
> }
> 
> // setWrapped implements the txnInterceptor interface.
> ```

This part of the interface is a little clunky, but I understand why you have it. So I'm not sure what I'm proposing. Nothing I guess?

---

*[pkg/kv/txn_span_refresher.go, line 47 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEy9EgW8ezPOdsgY-HD:-LEy9EgW8ezPOdsgY-HE:bzcp6qe) ([raw file](https://github.com/cockroachdb/cockroach/blob/c00da02874a2d2071a8f365974045fd9230f042c/pkg/kv/txn_span_refresher.go#L47)):*
> ```Go
> 	st *cluster.Settings
> 
> 	// mu locks the TxnCoordMeta.
> ```

This locks the meta, but `beforeSendLocked` also locks the meta. Are they (as in do they have to be for things to work) the same mutex? If so, point that out.

---

*[pkg/kv/txn_span_refresher.go, line 100 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEy8hFx3AKPSjwfo8jf:-LEy8hFx3AKPSjwfo8jg:b-q9itas) ([raw file](https://github.com/cockroachdb/cockroach/blob/c00da02874a2d2071a8f365974045fd9230f042c/pkg/kv/txn_span_refresher.go#L100)):*
> ```Go
> 	}
> 
> 	// If a prefix of the batch was executed, collect refresh spans for
> ```

Ah yeah, this hack bugged me before (returning a partial response in the presence of an error). We'll really have to soften the interface to `DistSender` somewhat. It's super subtle the way it works now.

---

*[pkg/kv/txn_span_refresher.go, line 157 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEy96Qy2-B38eQvWnaR:-LEy96Qy2-B38eQvWnaS:b-1k8c3h) ([raw file](https://github.com/cockroachdb/cockroach/blob/c00da02874a2d2071a8f365974045fd9230f042c/pkg/kv/txn_span_refresher.go#L157)):*
> ```Go
> // written more recently than the original transaction timestamp. All
> // implicated timestamp caches are updated with the final transaction
> // timestamp. On success, returns true and an updated BatchRequest
> ```

Does not return a `BatchRequest`.

---

*[pkg/kv/txn_span_refresher.go, line 163 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEy9RPe-BbI9zbHdq-h:-LEy9RPe-BbI9zbHdq-i:b17289h) ([raw file](https://github.com/cockroachdb/cockroach/blob/c00da02874a2d2071a8f365974045fd9230f042c/pkg/kv/txn_span_refresher.go#L163)):*
> ```Go
> 	ctx context.Context, refreshTxn *roachpb.Transaction,
> ) bool {
> 	sr.mu.Lock()
> ```

Struggling a bit to understand the locking here. Multiple clients are using this interceptor at the same time? So we have to guarantee that the spans here at least track this client's previous requests, which is trivially true, and the locking is only to avoid data races against other users (and we might pick up some of their spans but that's ok)?

---

*[pkg/kv/txn_span_refresher.go, line 177 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEyBWN3CJsK8joTwjqt:-LEyBWN3CJsK8joTwjqu:b152vf) ([raw file](https://github.com/cockroachdb/cockroach/blob/93303e5ae835bb45edf1d1310209c906c0f7ef52/pkg/kv/txn_span_refresher.go#L177)):*
> ```Go
> 	// Forward the refreshed timestamp under lock. This in conjunction with a
> 	// check in appendRefreshSpansLocked prevents a race where a concurrent
> 	// request may add new refresh spans only "verified" up to a certain after
> ```

a certain timestamp?

---

*[pkg/kv/txn_span_refresher.go, line 32 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEyUa663XKWL2wQ2ZFW:-LEyUa663XKWL2wQ2ZFX:b-e6piqq) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef672433a8a1af1767d681d699bd57b817f059a7/pkg/kv/txn_span_refresher.go#L32)):*
> ```Go
> const (
> 	// maxTxnRefreshAttempts defines the maximum number of times a single
> 	// transactional batch can trigger a refresh spans attempt.
> ```

Mind adding why an attempt would fail?

---

*[pkg/kv/txn_span_refresher.go, line 91 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEyUeOsENULRtsOsPfF:-LEyUeOsENULRtsOsPfG:bs4q9op) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef672433a8a1af1767d681d699bd57b817f059a7/pkg/kv/txn_span_refresher.go#L91)):*
> ```Go
> 	}
> 
> 	// Send through wrapped lockedSender. Unlocks while sending then re-locks.
> ```

Ditto.

---

*[pkg/kv/txn_span_refresher.go, line 164 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEyVTv_AQSFLa6i9hUZ:-LEyVTv_AQSFLa6i9hU_:b-2o2lrg) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef672433a8a1af1767d681d699bd57b817f059a7/pkg/kv/txn_span_refresher.go#L164)):*
> ```Go
> 	// then combines the results, partial success makes it very
> 	// difficult to determine what can be retried.
> 	if aPSErr, ok := pErr.GetDetail().(*roachpb.MixedSuccessError); ok {
> ```

Not new code, but this seems brittle. I really hope these distant pieces all work together as they ought to.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-:-LEyW1H05w9nkHjVVwt2:b-r0d12p)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 5 files at r4.
Review status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

---

*[pkg/kv/txn_coord_sender.go, line 164 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEkckJR1kkBI8FtRpay:-LEkckJR1kkBI8FtRpaz:b-sm4xum) ([raw file](https://github.com/cockroachdb/cockroach/blob/976264fc5b340677021d386c2b952188815f863b/pkg/kv/txn_coord_sender.go#L164)):*
> ```Go
> 	populateMetaLocked(meta *roachpb.TxnCoordMeta)
> 
> 	// refreshMetaLocked refreshes any internal state held inside the
> ```

comment is out of sync

---

*[pkg/kv/txn_coord_sender.go, line 171 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEkcq-lBucHxEiZ_9e5:-LEkcq-lBucHxEiZ_9e6:b986py4) ([raw file](https://github.com/cockroachdb/cockroach/blob/976264fc5b340677021d386c2b952188815f863b/pkg/kv/txn_coord_sender.go#L171)):*
> ```Go
> 	// epochRetryLocked resets the interceptor in the case of a txn epoch
> 	// increment.
> 	epochRetryLocked()
> ```

nit: maybe `newEpochLocked` or `epochBumpedLocked`

---

*[pkg/kv/txn_coord_sender.go, line 161 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEzghzdE_TXIp0Xf7Bp:-LEzghzdE_TXIp0Xf7Bq:b-92m68x) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef672433a8a1af1767d681d699bd57b817f059a7/pkg/kv/txn_coord_sender.go#L161)):*
> ```Go
> var _ client.TxnSender = &TxnCoordSender{}
> 
> // lockedSender is like a client.Sender but requires the called to hold the
> ```

the caller

---

*[pkg/kv/txn_coord_sender.go, line 163 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEzgjQn8HUMulvuUZXN:-LEzgjQn8HUMulvuUZXO:b-yzbn57) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef672433a8a1af1767d681d699bd57b817f059a7/pkg/kv/txn_coord_sender.go#L163)):*
> ```Go
> // lockedSender is like a client.Sender but requires the called to hold the
> // TxnCoordSender lock to send requests.
> type lockedSender interface {
> ```

I think it'd be good if this comment had more words. Explain when the lock is acquired and by who and when released.
Also say why - say that all the state of all the implemeters is updated atomically in two pahses - outgoing and incoming.

---

*[pkg/kv/txn_intent_collector.go, line 152 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEkdlOR9gdq_V8TBhUV:-LEkdlOR9gdq_V8TBhUW:bhfud5m) ([raw file](https://github.com/cockroachdb/cockroach/blob/976264fc5b340677021d386c2b952188815f863b/pkg/kv/txn_intent_collector.go#L152)):*
> ```Go
> // augmentMetaLocked implements the txnReqInterceptor interface.
> func (ic *txnIntentCollector) augmentMetaLocked(meta roachpb.TxnCoordMeta) {
> 	// Do not modify existing intent slices when copying.
> ```

I'm not sure I understand this comment.  Why do we copy `ic.intents`?

---

*[pkg/kv/txn_intent_collector.go, line 126 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEyUQYe2iKY1XbJL4SY:-LEzn663CVVTAS6cllh3:b59b1lk) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef672433a8a1af1767d681d699bd57b817f059a7/pkg/kv/txn_intent_collector.go#L126)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

This part of the interface is a little clunky, but I understand why you have it. So I'm not sure what I'm proposing. Nothing I guess?
</blockquote></details>

it also bothers me. I think what we can do is have the TCS hook all the intercepters together without going through this interface. It knows all their types in the alloc.

---

*[pkg/kv/txn_span_refresher.go, line 290 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEzq-wR0-3fH2PnjRON:-LEzq-wSArGX0sYLXWRX:b-sr6kvo) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef672433a8a1af1767d681d699bd57b817f059a7/pkg/kv/txn_span_refresher.go#L290)):*
> ```Go
> //
> // The batch refreshed timestamp and the max refreshed timestamp for the sender
> // can get out of step because the txn coord sender can be used concurrently
> ```

s/txn coord sender/txnSpanRefresher

---

*[pkg/sql/txn_restart_test.go, line 1356 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEzu0GY1D0CNAb4ocTu:-LEzu0GY1D0CNAb4ocTv:b-65l5i6) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef672433a8a1af1767d681d699bd57b817f059a7/pkg/sql/txn_restart_test.go#L1356)):*
> ```Go
> 			// Allow six restarts so that the auto retry on the first few
> 			// uncertainty interval errors also fails.
> 			// WIP: this should be parameterized somehow and hooked up to
> ```

don't forget

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-:-LEzuP4h6yzfm0dR1JlN:bnfp4nl)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
LGTM. It seems like a lot of abstraction and complexity for a small, fixed number of interceptors (especially if the batch wrapper goes away). But I defer to y'all on whether this is a good move. 

---

Reviewed 1 of 5 files at r1, 1 of 5 files at r8.
Review status: :shipit: complete! 1 of 0 LGTMs obtained

---

*[pkg/kv/txn_batch_wrapper.go, line 29 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LF3c9qpDaT0MyLF767n:-LF3c9qpDaT0MyLF767o:b-fwfq3r) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef672433a8a1af1767d681d699bd57b817f059a7/pkg/kv/txn_batch_wrapper.go#L29)):*
> ```Go
> //
> // TODO(tschottdorf): this handling is somewhat awkward but unless we want to
> // give this error back to the client, our options are limited. We'll have to
> ```

As we've discussed offline, I'd like to see this behavior go away - return the error to the caller and require all callers that expect atomicity to use an explicit transaction. (I think that's safe - our non-transactional requests are nearly always point operations outside of tests).

---

*[pkg/kv/txn_coord_sender.go, line 213 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEyTgyU-gU3bcLdU8ei:-LF3ZgZr5JXU34nAtfi5:b-rq9cmf) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef672433a8a1af1767d681d699bd57b817f059a7/pkg/kv/txn_coord_sender.go#L213)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Sneaky, I like it. But you could put a comment here.
</blockquote></details>

Yeah, this needs a loud warning comment. A "locked" method that releases the lock is unexpected. 

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-:-LF3lL646lg0FAxpOw42:bnxnkwf)*
<!-- Sent from Reviewable.io -->

nvb committed:
kv: create ClientTestingKnobs and TxnCoordSenderFactoryConfig

This change makes two related changed. First, it pulls all of the
auxiliary objects from `TxnCoordSenderFactory`'s constructor into
a new TxnCoordSenderFactoryConfig struct. This allows them to be
optional.

Next, it creates a new ClientTestingKnobs struct that subsumes the
DistSenderTestingKnobs struct and contains a new `MaxTxnRefreshAttempts`
field. This allows us to clean up `TestReacquireLeaseOnRestart`.

Release note: None

nvb committed:
kv: rename txnInterceptor files

Release note: None

nvanbenschoten commented:
>  It seems like a lot of abstraction and complexity for a small, fixed number of interceptors (especially if the batch wrapper goes away)

The batch wrapper will be going away soon, but I already have plans to add two other interceptors in its place. This framework has made them a lot easier to develop in isolation without needing to make invasive changes to the `TxnCoordSender`.

---

Review status: :shipit: complete! 1 of 0 LGTMs obtained

---

*[pkg/kv/txn_batch_wrapper.go, line 1 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEy9xZ16KtcN-Hhmkc0:-LFJRTHk6l1rv2cKTKok:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c00da02874a2d2071a8f365974045fd9230f042c/pkg/kv/txn_batch_wrapper.go#L1)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Maybe the scheme `(txn_)?interceptor_<name>.go` is more appropriate. Your call.
</blockquote></details>

Done.

---

*[pkg/kv/txn_batch_wrapper.go, line 29 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LF3c9qpDaT0MyLF767n:-LFIWv9VBCIEKMck6nR2:bndr6pb) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef672433a8a1af1767d681d699bd57b817f059a7/pkg/kv/txn_batch_wrapper.go#L29)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

As we've discussed offline, I'd like to see this behavior go away - return the error to the caller and require all callers that expect atomicity to use an explicit transaction. (I think that's safe - our non-transactional requests are nearly always point operations outside of tests).
</blockquote></details>

#26741. Thanks Andrei!

---

*[pkg/kv/txn_coord_sender.go, line 164 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEkckJR1kkBI8FtRpay:-LFIYCsaAt1xTQm2P-tL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/976264fc5b340677021d386c2b952188815f863b/pkg/kv/txn_coord_sender.go#L164)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

comment is out of sync
</blockquote></details>

Done.

---

*[pkg/kv/txn_coord_sender.go, line 171 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEkcq-lBucHxEiZ_9e5:-LFIYRoA81IaRmwl9UYe:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/976264fc5b340677021d386c2b952188815f863b/pkg/kv/txn_coord_sender.go#L171)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

nit: maybe `newEpochLocked` or `epochBumpedLocked`
</blockquote></details>

Done.

---

*[pkg/kv/txn_coord_sender.go, line 171 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEy3b6UDxPGAUTjos-T:-LFIYG2_9n2FJVBMf5zp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c00da02874a2d2071a8f365974045fd9230f042c/pkg/kv/txn_coord_sender.go#L171)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

"when" -> before or after?
</blockquote></details>

Done.

---

*[pkg/kv/txn_coord_sender.go, line 491 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEy4COI7CfiIxYp6ILB:-LFI__so1FDoRad_wYv0:by9jpa1) ([raw file](https://github.com/cockroachdb/cockroach/blob/c00da02874a2d2071a8f365974045fd9230f042c/pkg/kv/txn_coord_sender.go#L491)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

why return `ba` on error here and above? Either something subtle is happening or it should be changed. But it looks like the caller just throws `ba` away on error.
</blockquote></details>

I think because it was a value and not a pointer, but this is gone now anyway.

---

*[pkg/kv/txn_coord_sender.go, line 518 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEy4_6e8Q4evqEXiyWd:-LFI_gfB0nSFxkyLv1we:b53yq7j) ([raw file](https://github.com/cockroachdb/cockroach/blob/c00da02874a2d2071a8f365974045fd9230f042c/pkg/kv/txn_coord_sender.go#L518)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

The abstraction is a little leaky here, right? On the way out, the interceptor sees the request before any later interceptors made their modifications. Then, on the way back, it gets a `ba` that has seen all of the modifications and in particular may have been altered by all later senders both on the way out and in.

Not necessarily a problem, but call this out in the command on `maybeRetrySend`.
</blockquote></details>

Yes exactly, and this is how this `maybeRetrySend` has always worked. In fact, without this, the refresh spans optimization would not work at all.

I removed this all in the last commit and made the updated state more explicit. See `largestRefreshTS` in `txnSpanRefresher`.

---

*[pkg/kv/txn_coord_sender.go, line 186 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEyAn646UtjckUjk4lt:-LFIYC_A0KOSXuMb6SEE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/93303e5ae835bb45edf1d1310209c906c0f7ef52/pkg/kv/txn_coord_sender.go#L186)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Stale
</blockquote></details>

Done.

---

*[pkg/kv/txn_coord_sender.go, line 334 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEyTmuG8BIvJ-6pFfIK:-LFI_WdQ6axJNLyqwHeR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/93303e5ae835bb45edf1d1310209c906c0f7ef52/pkg/kv/txn_coord_sender.go#L334)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

:thumbsup: there goes my earlier comment.
</blockquote></details>

Done.

---

*[pkg/kv/txn_coord_sender.go, line 161 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEzghzdE_TXIp0Xf7Bp:-LFIX1G6-hTIl8oOEO2r:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef672433a8a1af1767d681d699bd57b817f059a7/pkg/kv/txn_coord_sender.go#L161)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

the caller
</blockquote></details>

Done.

---

*[pkg/kv/txn_coord_sender.go, line 163 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEzgjQn8HUMulvuUZXN:-LFIXu9N8P9BMIKnirOR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef672433a8a1af1767d681d699bd57b817f059a7/pkg/kv/txn_coord_sender.go#L163)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I think it'd be good if this comment had more words. Explain when the lock is acquired and by who and when released.
Also say why - say that all the state of all the implemeters is updated atomically in two pahses - outgoing and incoming.

</blockquote></details>

Done.

---

*[pkg/kv/txn_coord_sender.go, line 213 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEyTgyU-gU3bcLdU8ei:-LFIYimjDPhNwzOPOo5G:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef672433a8a1af1767d681d699bd57b817f059a7/pkg/kv/txn_coord_sender.go#L213)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Yeah, this needs a loud warning comment. A "locked" method that releases the lock is unexpected. 
</blockquote></details>

Done.

---

*[pkg/kv/txn_intent_collector.go, line 152 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEkdlOR9gdq_V8TBhUV:-LFJ2LHB4JrKh0InATks:b-v7r35b) ([raw file](https://github.com/cockroachdb/cockroach/blob/976264fc5b340677021d386c2b952188815f863b/pkg/kv/txn_intent_collector.go#L152)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I'm not sure I understand this comment.  Why do we copy `ic.intents`?
</blockquote></details>

I don't think it's necessary, but I think it's a good idea to copy under lock instead of sharing access to the slice. This isn't new.

---

*[pkg/kv/txn_intent_collector.go, line 72 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEy649u2bEHcRB2ng_Z:-LFIaAym6GfK8kxyVK8D:b-463r0r) ([raw file](https://github.com/cockroachdb/cockroach/blob/c00da02874a2d2071a8f365974045fd9230f042c/pkg/kv/txn_intent_collector.go#L72)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

I wouldn't mind if this TODO had a ... happy little accident :gun: 
</blockquote></details>

Whoops... the TODO must have been lost during the code movement :see_no_evil: 

---

*[pkg/kv/txn_intent_collector.go, line 133 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEy7ZYg5asXjxjZnAED:-LFJ7dKM5g2vZgwdBnkY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c00da02874a2d2071a8f365974045fd9230f042c/pkg/kv/txn_intent_collector.go#L133)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

I haven't peeked inside of this method, but it seems pretty bad if we're sometimes missing intent spans, for that could lead to lost writes if the txn ends up committing.
</blockquote></details>

Done.

---

*[pkg/kv/txn_intent_collector.go, line 140 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEy7jYdEdjhYmu54R7a:-LFJ2G-1CLhAKnimC9x-:b6qtsd5) ([raw file](https://github.com/cockroachdb/cockroach/blob/c00da02874a2d2071a8f365974045fd9230f042c/pkg/kv/txn_intent_collector.go#L140)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Maybe this method should be called `metaRefreshedLocked()`?
</blockquote></details>

This is gone in later commits.

---

*[pkg/kv/txn_intent_collector.go, line 157 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEy7ySz8m0BxyO66Lck:-LFJ2TqNEr_yrwzMLHXQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c00da02874a2d2071a8f365974045fd9230f042c/pkg/kv/txn_intent_collector.go#L157)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

The mystery continues.
</blockquote></details>

Done.

---

*[pkg/kv/txn_intent_collector.go, line 177 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEy86NqBpyE8aA3hCM6:-LFJ4ctM8esXRoVjb2ry:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c00da02874a2d2071a8f365974045fd9230f042c/pkg/kv/txn_intent_collector.go#L177)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Aha! I don't think that should return an error. It should just leave the spans uncondensed.
</blockquote></details>

Done.

---

*[pkg/kv/txn_intent_collector.go, line 116 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEyUKQE4xfcacFZcE-Z:-LFJ6FoGEkdz5bpFNb46:bth9biy) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef672433a8a1af1767d681d699bd57b817f059a7/pkg/kv/txn_intent_collector.go#L116)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

What's the comment for? This method doesn't care any more, right?
</blockquote></details>

Well I wanted to point out that the lock isn't held across this call.

---

*[pkg/kv/txn_intent_collector.go, line 126 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEyUQYe2iKY1XbJL4SY:-LFJ1qXtCeA8hTq3ghtO:b-c9p2z1) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef672433a8a1af1767d681d699bd57b817f059a7/pkg/kv/txn_intent_collector.go#L126)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

it also bothers me. I think what we can do is have the TCS hook all the intercepters together without going through this interface. It knows all their types in the alloc.
</blockquote></details>

Yeah. we could have the TCS hook them all together, but we'd then have two places where we defined the order of the interceptor stack. I don't like this either, but I prefer it to duplicated logic that could get out of sync.

---

*[pkg/kv/txn_span_refresher.go, line 47 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEy9EgW8ezPOdsgY-HD:-LFJ68oKCeH5kXQEjH0M:bss4a2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/c00da02874a2d2071a8f365974045fd9230f042c/pkg/kv/txn_span_refresher.go#L47)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

This locks the meta, but `beforeSendLocked` also locks the meta. Are they (as in do they have to be for things to work) the same mutex? If so, point that out.
</blockquote></details>

Removed.

---

*[pkg/kv/txn_span_refresher.go, line 157 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEy96Qy2-B38eQvWnaR:-LFJ6jat3IscOKM5FQTK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c00da02874a2d2071a8f365974045fd9230f042c/pkg/kv/txn_span_refresher.go#L157)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Does not return a `BatchRequest`.
</blockquote></details>

Done.

---

*[pkg/kv/txn_span_refresher.go, line 177 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEyBWN3CJsK8joTwjqt:-LFJ7DE_1vlcI7_5OyP8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/93303e5ae835bb45edf1d1310209c906c0f7ef52/pkg/kv/txn_span_refresher.go#L177)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

a certain timestamp?
</blockquote></details>

Done.

---

*[pkg/kv/txn_span_refresher.go, line 32 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEyUa663XKWL2wQ2ZFW:-LFJ67w48pwaDm49wlxl:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef672433a8a1af1767d681d699bd57b817f059a7/pkg/kv/txn_span_refresher.go#L32)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Mind adding why an attempt would fail?
</blockquote></details>

Done.

---

*[pkg/kv/txn_span_refresher.go, line 290 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEzq-wR0-3fH2PnjRON:-LFJ7HlkCyhgK6kD-713:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef672433a8a1af1767d681d699bd57b817f059a7/pkg/kv/txn_span_refresher.go#L290)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

s/txn coord sender/txnSpanRefresher
</blockquote></details>

Done.

---

*[pkg/sql/txn_restart_test.go, line 1356 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LEzu0GY1D0CNAb4ocTu:-LFJROVV-rdIVAnALKHI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef672433a8a1af1767d681d699bd57b817f059a7/pkg/sql/txn_restart_test.go#L1356)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

don't forget
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-:-LFJ7MtG98gRH1EkQBCr:b-ojtj2g)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

---

*[pkg/base/testing_knobs.go, line 27 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LFJVRGjC_Xa0Fy6C61b:-LFJVRGjC_Xa0Fy6C61c:b-723e71) ([raw file](https://github.com/cockroachdb/cockroach/blob/80b99c0b3e71d4f7a2fa922cec55badfa5e702a7/pkg/base/testing_knobs.go#L27)):*
> ```Go
> type TestingKnobs struct {
> 	Store               ModuleTestingKnobs
> 	KVClient            ModuleTestingKnobs
> ```

Maybe just `Client` here for consistency with the type name

---

*[pkg/kv/testing_knobs.go, line 19 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LFJW-Qy4kTMdP_JY0hv:-LFJW-Qy4kTMdP_JY0hw:b-11l4jo) ([raw file](https://github.com/cockroachdb/cockroach/blob/80b99c0b3e71d4f7a2fa922cec55badfa5e702a7/pkg/kv/testing_knobs.go#L19)):*
> ```Go
> import "github.com/cockroachdb/cockroach/pkg/base"
> 
> // ClientTestingKnobs is a part of the context used to control parts of
> ```

nit: The word "context" here has became archaic

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-:-LFJWdYu0J8uhAfE6c30:bnfp4nl)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
TFTR.

---

Review status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

---

*[pkg/base/testing_knobs.go, line 27 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LFJVRGjC_Xa0Fy6C61b:-LFJY4UL08qxmxKnanI7:b-fcde0z) ([raw file](https://github.com/cockroachdb/cockroach/blob/80b99c0b3e71d4f7a2fa922cec55badfa5e702a7/pkg/base/testing_knobs.go#L27)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Maybe just `Client` here for consistency with the type name
</blockquote></details>

I went back and forth on this but decided on KVClient because it doesn't leave room for ambiguity.

---

*[pkg/kv/testing_knobs.go, line 19 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-LFJW-Qy4kTMdP_JY0hv:-LFJYcFpASRtRJvuXC81:b-qu6n94) ([raw file](https://github.com/cockroachdb/cockroach/blob/80b99c0b3e71d4f7a2fa922cec55badfa5e702a7/pkg/kv/testing_knobs.go#L19)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

nit: The word "context" here has became archaic
</blockquote></details>

Good point. Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26496#-:-LFJY3G1D0xmdB7prQqo:b-814g87)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
bors r+
